serialize/decoder.go

```diff
@@ -0,0 +1,200 @@
+	"github.com/m3db/m3x/ident"
+)
+
+var (
+	errIncorrectHeader               = errors.New("header magic number does not match expected value")
+	errInvalidByteStreamIDDecoding   = errors.New("internal error, invalid byte stream while decoding ID")
+	errInvalidByteStreamUintDecoding = errors.New("internal error, invalid byte stream while decoding uint")
+)
+
+type decoder struct {
+	checkedData checked.Bytes
+	data        []byte
+	remaining   int
+	err         error
+	hasCurrent  bool
+	current     ident.Tag
+
+	pool        DecoderPool
+	wrapperPool xpool.CheckedBytesWrapperPool
+}
+
+func newDecoder(pool DecoderPool, wrapperPool xpool.CheckedBytesWrapperPool) Decoder {
+	return &decoder{
+		pool:        pool,
+		wrapperPool: wrapperPool,
+	}
+}
+
+func (d *decoder) Reset(b checked.Bytes) {
+	d.checkedData = b
+	d.checkedData.IncRef()
+	d.data = d.checkedData.Get()
+
+	header, err := d.decodeUInt16()
+	if err != nil {
+		d.err = err
+		return
+	}
+
+	if header != HeaderMagicNumber {
+		d.err = errIncorrectHeader
+		return
+	}
+
+	remain, err := d.decodeUInt16()
+	if err != nil {
+		d.err = err
+		return
+	}
+
+	d.remaining = int(remain)
+}
+
+func (d *decoder) Next() bool {
+	if d.err != nil || d.remaining <= 0 {
+		return false
+	}
+
+	d.release()
+	t, err := d.decodeTag()
+	if err != nil {
+		d.err = err
+		return false
+	}
+
+	d.remaining--
+	d.current = t
+	d.hasCurrent = true
+	return true
+}
+
+func (d *decoder) Current() ident.Tag {
+	return d.current
+}
+
+func (d *decoder) decodeTag() (ident.Tag, error) {
+	name, err := d.decodeID()
+	if err != nil {
+		return ident.Tag{}, err
+	}
+
+	value, err := d.decodeID()
+	if err != nil {
+		name.Finalize()
+		return ident.Tag{}, err
+	}
+
+	return ident.Tag{name, value}, nil
+}
+
+func (d *decoder) decodeID() (ident.ID, error) {
+	l, err := d.decodeUInt16()
+	if err != nil {
+		return nil, err
+	}
+
+	if len(d.data) < int(l) {
+		return nil, errInvalidByteStreamIDDecoding
+	}
+
+	// incRef to indicate another checked.Bytes has a
+	// reference to the original bytes
+	d.checkedData.IncRef()
+	b := d.wrapperPool.Get(d.data[:l])
+	d.data = d.data[l:]
+
+	return ident.BinaryID(b), nil
+}
+
+func (d *decoder) decodeUInt16() (uint16, error) {
+	if len(d.data) < 2 {
+		return 0, errInvalidByteStreamUintDecoding
+	}
+
+	n := decodeUInt16(d.data)
+	d.data = d.data[2:]
+	return n, nil
+}
+
+func (d *decoder) Err() error {
+	return d.err
+}
+
+func (d *decoder) release() {
+	if n := d.current.Name; n != nil {
+		n.Finalize()
+		d.checkedData.DecRef() // indicate we've released the extra ref
+		d.current.Name = nil
+	}
+	if v := d.current.Value; v != nil {
+		v.Finalize()
+		d.checkedData.DecRef() // indicate we've released the extra ref
+		d.current.Value = nil
+	}
+}
+
+func (d *decoder) Remaining() int {
+	return d.remaining
+}
+
+func (d *decoder) Close() {
+	d.release()
+	d.data = nil
+	d.err = nil
+	d.remaining = 0
+	if d.checkedData != nil {
+		d.checkedData.DecRef()
+		d.checkedData = nil
```

Comment:
Should also `d.checkedData.Finalize()` after the decref so it get's returned to pool.

Reply:
Hm guess you'll need to know if this one got duplicated, and if so, not to finalize I guess.

Reply:
Good call. I can check `NumRef()==0` and Finalize() based on that. Sound good?

Reply:
done